c/src/atom.rs

```diff
@@ -171,6 +196,106 @@ pub extern "C" fn bindings_resolve_and_remove(bindings: *mut bindings_t, var_nam
 
 // end of bindings
 
+// bindings_set
+
+#[no_mangle]
+pub extern "C" fn bindings_set_empty() -> *mut bindings_set_t {
+    bindings_set_into_ptr(BindingsSet::empty())
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_single() -> *mut bindings_set_t {
+    bindings_set_into_ptr(BindingsSet::single())
+}
+
+/// WARNING: This function takes ownership of the bindings argument.
+/// After calling this function, the bindings_t passed must not be accessed or freed
+/// 
+/// The object returned from this function must be freed with bindings_set_free()
+#[no_mangle]
+pub extern "C" fn bindings_set_from_bindings(bindings: *mut bindings_t) -> *mut bindings_set_t {
+    let owned_bindings = ptr_into_bindings(bindings);
+    bindings_set_into_ptr(BindingsSet::from(owned_bindings))
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_free(set: *mut bindings_set_t) {
+    // drop() does nothing actually, but it is used here for clarity
+    drop(unsafe{Box::from_raw(set)});
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_clone(set: *const bindings_set_t) -> *mut bindings_set_t {
+    let set = unsafe{&(*set).set};
+    bindings_set_into_ptr(set.clone())
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_eq(set: *const bindings_set_t, other: *const bindings_set_t) -> bool {    
+    let set = unsafe{&(*set).set};
+    let other = unsafe{&(*other).set};
+    set == other
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_to_str(set: *const bindings_set_t, callback: c_str_callback_t, context: *mut c_void) {
+    let set = unsafe{ &(*set).set };
+    let s = str_as_cstr(set.to_string().as_str());
+    callback(s.as_ptr(), context);
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_len(set: *const bindings_set_t) -> usize {
+    let set = unsafe{ &(*set).set };
+    set.len()
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_iterate(set: *const bindings_set_t, callback: bindings_callback_t, context: *mut c_void) {
+    let set = unsafe{ &(*set).set };
+    for bindings in set.iter() {
+        let bindings_ptr = (bindings as *const Bindings).cast::<bindings_t>();
+        callback(bindings_ptr, context);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_add_var_equality(set: *mut bindings_set_t, a: *const atom_t, b: *const atom_t) {
+    let set = unsafe{ &mut(*set).set };
+    let a = unsafe{ &(*a).atom };
+    let b = unsafe{ &(*b).atom };
+
+    let mut owned_set = BindingsSet::empty();
+    core::mem::swap(&mut owned_set, set);
+    let mut result_set = owned_set.add_var_equality(a.try_into().unwrap(), b.try_into().unwrap());
+    core::mem::swap(&mut result_set, set);
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_set_add_var_binding(set: *mut bindings_set_t, var: *const atom_t, value: *const atom_t) {
```

Comment:
Playing with API I have found an issue. If one uses `BindingsSet::empty()` to create new `BindingsSet` instance then it is impossible to add something into it. It always ends up with an empty `BindingsSet` instance.

Reply:
Good point.  I actually can't think of any good reason to expose the BindingsSet::empty() constructor to the user.  As you point out, in light of the other functionality we've chosen to expose, BindingsSet::empty is actually useless.  Maybe this should be a private method in Rust, and not exposed to C & Python.

Reply:
I removed if from the C & Python APIs.  I left it in Rust because it provided a handy way to make an object to provide to mem::swap() to take ownership from a mutable pointer to a BindingsSet in the C API.

I want to find a better way to describe the meaning of the different states of a BindingsSet; and when we agree on something I will harmonize the code.

Reply:
Thanks to your comment, I now see the value in keeping the empty constructor, for composing the return value from a grounded atom match.  So I've re-added it.